### DISCUSSION TITLE
BTP - Code Structure Refactor

### DISCUSSION THREAD

**Discussion by rise-techiast** [icon-project/btp#25] [Q&A] [answered]
Dear teams,

To follow up with this layout structure recommendation https://iconcommunity.slack.com/archives/C01TC9TRE74/p1617873261027200, we would like to propose a structure that arose from our code discovery & refactor. We published this branch https://github.com/icon-project/btp/tree/btp-layout-refactor to compare with out current master, please review and let us know if we should proceed with an official PR.

What we have done:

- Separated the core BMR in a new folder.
- Simplify docker environment.
- Support hot reload.

Pros:

- Easier for reading code.
- Easier for development & scalability.

Cons:

- The new layout might not be familiar with who have maintained from the old layout.

Sincerely,

**Comment by cow-hs** [comment]
Hi.
First of all, thank you for your suggestion.

However, more discussion is needed on accepting the offer.

Questions)
1. Is it a story about dividing the contents of btpsimple/chain/chain.go into three such as bmr.go, bmr_internal.go, and bmr_rotate 
Is there more than this?

It would be nice if you could give more examples to say that development and scalability are easier.

**Reply by canhlinh** [reply]
@leeheonseung 

The main reason is that we want a cleaner code layout so that we can see what is going on inside easier. (More readable). I would like to show a few points:

1. Moving [all core logic ](https://github.com/icon-project/btp/tree/master/cmd/btpsimple/chain) to the [btp](https://github.com/icon-project/btp/tree/refactor-layout/btp) package and rewriting code to make it easier to read -> Developers will be aware of where the main logic of BTP is.

2. Moving [module](https://github.com/icon-project/btp/tree/master/cmd/btpsimple/module) to [chain](https://github.com/icon-project/btp/tree/refactor-layout/chain). Here you can find out which chains we support. `chain` sound a better name right?

3. All command lines have been moved to the [cmd package](https://github.com/icon-project/btp/tree/refactor-layout/cmd/btpsimple). Only logic related to routing execution should be kept in this package. We also have spirit 4 code files here for easy reading.

3. Moving [docker-compose](https://github.com/icon-project/btp/tree/master/docker-compose) to [testnet](https://github.com/icon-project/btp/tree/refactor-layout/testnet) We might have more docker-compose files in this folder. Each file is an illusion of integrating between 2 chains. Such as ICX -> Parachain or ICX->ICX...  

4. In the main folder, make a [Dockerfile](https://github.com/icon-project/btp/blob/refactor-layout/Dockerfile). I simply want to run BTP code in a Docker container. Dev can function without the installation of Golang. The Dockerfile also allows for auto-compilation upon save. It's very convenient for development. Do you like this function?

What do you think?

**Comment by trmaphi** [comment, marked as answer]
As confirmed on Slack, hugobytes and web3labs already agrees with this change.
